File: scanner/module/dns/axfr.go
package dns

import (
	"net"
	"os/exec"
	"strings"

	"github.com/bloom42/phaser/common/phaser"
	"github.com/bloom42/phaser/common/phaser/findings"
	"github.com/bloom42/phaser/scanner/module"
)

type ZoneTransferInformationDisclosure struct{}

func (ZoneTransferInformationDisclosure) Name() string {
	return "dns/zone_trasnfer_information_disclosure"
}

func (ZoneTransferInformationDisclosure) Description() string {
	return "check if AXFR queries are enabled for each name server"
}

func (ZoneTransferInformationDisclosure) Author() string {
	return "Sylvain Kerkour <sylvain@kerkour.com>"
}

func (ZoneTransferInformationDisclosure) Version() string {
	return "0.1.0"
}

func (ZoneTransferInformationDisclosure) Run(scan *phaser.Scan, target *phaser.Target) (module.Result, []error) {
	errs := []error{}
	badServers := []findings.AXFR{}
	var ret module.Result

	nservers, err := net.LookupNS(target.Host)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}

	for _, ns := range nservers {
		command := "dig"
		commandArgs := []string{"AXFR", target.Host, "@" + ns.Host}

		out, err := exec.Command(command, commandArgs...).Output()
		if err != nil {
			errs = append(errs, err)
			// var errr string
			// if eerr, ok := err.(*exec.ExitError); ok {
			// 	errr = string(eerr.Stderr)
			// } else {
			// 	errr = err.Error()
			// }
			// ret.Errors = append(ret.Errors, formatError(errr, target.Host, target.Type, nil))
			continue
		}

		outStr := string(out)
		outStrLower := strings.ToLower(outStr)
		if strings.Contains(outStr, "XFR") && strings.Contains(outStrLower, "transfer failed") == false {
			badServer := findings.AXFR{
				Domain:   ns.Host,
				Response: outStr,
			}
			badServers = append(badServers, badServer)
		}
	}

	if len(badServers) != 0 {
		ret = badServers
	}
	return ret, errs
}

File: src/scanner/modules/domain/axfr.rs
use crate::{
    error::PhaserError,
    scanner::{
        module,
        findings,
        Scan,
        Target,
        TargetKind,
    },
};
use std::process::{Command};

pub struct Axfr{}

impl module::BaseModule for Axfr {
    fn name(&self) -> String {
        return String::from("domain/axfr");
    }

    fn description(&self) -> String {
        return String::from("check if AXFR queries are enabled for each name server");
    }

    fn author(&self) -> String {
        return String::from("Sylvain Kerkour <sylvain@kerkour.com>")
    }

    fn version(&self) -> String {
        return String::from("0.1.0");
    }
}

impl module::HostModule for Axfr {
    fn run(&self, _: &Scan, target: &Target) -> Result<findings::Data, PhaserError> {
        let mut data = vec!();

        if let TargetKind::Ip = target.kind {
            return Ok(findings::Data::None);
        };

        // first retrieve NS servers
        let dig_output = Command::new("dig")
            .arg("+short")
            .arg("NS")
            .arg(&target.host)
            .output()?;
        let ns_output = String::from_utf8_lossy(&dig_output.stdout).to_string();

        if ns_output.is_empty() {
            return Ok(findings::Data::None);
        }

        let ns_servers: Vec<&str> = ns_output.split(',').collect();

        // for each server, try AXFR DNS query
        for ns_server in ns_servers {
            let mut output = String::new();
            match Command::new("dig")
                .arg("AXFR")
                .arg(&target.host)
                .arg(format!("@{}", ns_server))
                .output()
                {
                Ok(dig_output) => output = String::from_utf8_lossy(&dig_output.stdout).to_string(),
                Err(_) => {},
            };
            if output.contains("XFR") && !output.contains("transfer failed") {
                data.push(findings::domain::Axfr{
                    server: ns_server.to_string(),
                    response: output,
                });
            }
        }

        if data.len() != 0 {
            return Ok(findings::Data::Axfr(data));
        }

        return Ok(findings::Data::None);
    }
}


#[cfg(test)]
mod tests {
    use crate::scanner::module::BaseModule;

    #[test]
    fn module_name() {
        let module = super::Axfr{};
        assert_eq!("domain/axfr", module.name());
    }
}
